### ISSUE TITLE
documentation: examples should be more practical

### ISSUE THREAD

**Issue by AdrianRossouw** [nickfargo/state#18]
I'm talking of course about things like the bleep bloop example, where the actual operation simply has no ground in reality. You have to really get invested in the code to comprehend not only what it is doing, but what it is intended to do.

What this also means I think is that it's very difficult to understand why you would use this library in the first place. I don't know if I would go as far as writing a todoMVC, but I definitely think there needs to be a more 'state.js for humans' side to the documentation.

**Comment by nickfargo** [comment]
Yeah, a few of the examples are utter napkin-scribbles, abstract and contrived, to put it kindly. They’re at their worst toward the bottom of the docs page; [“Transition expressions”](http://statejs.org/docs/#concepts--transitions--expressions) and [“State guards”](http://statejs.org/docs/#concepts--guards--state-guards) are particularly egregious.

Ideal would be examples that are at once practical, illustrative, and concise — though that’s an elusive trifecta at my current example-writing skill level — and that shiny ideal usually ends up devolving into one of those “choose any two” conundrums.

Anyway, next best thing, likely, is multiple practical examples for each section of the docs, which individually may utilize only a fraction of a particular feature, but collectively would cover all or most of the core principles.

And of course, pull requests always welcome on gh-pages too … ;)

**Comment by AdrianRossouw** [comment]
Well, mulling the [todoMVC](todomvc.com) thing over it wouldn't be the most terrible idea. It might be bland, but it's a widely used idiom in the developer community, and it's certainly a step up from a widget factory.

Tasks have rather inherent and explicit states (in progress, completed). There is opportunity to explain guards (can't add a task with no title), there is opportunity to explain events and transitions (interface needs to redraw itself when you complete a task).

It would also tackle some of the things that I found real pain points, such as persisting and re-inflating in progress state machines. It also has a pre-existing codebase that can be forked from to build the examples : [vanillajs-examples](https://github.com/tastejs/todomvc/tree/gh-pages/vanilla-examples/vanillajs)

I'm also going to venture that the reason you are having trouble finding the right examples is probably because you are suffering from a pretty pronounced case of the [curse of knowledge](http://en.wikipedia.org/wiki/Curse_of_knowledge). I think the problem at hand pretty much necessitates a fresh set of eyes, and I'm willing to help.

As with all open source contributions though, I can't commit to a schedule.. but I am going to need to document the operation of state.js for the developers on the projects I am using it for anyway, so it is really in my best interest to contribute this via proper channels

**Comment by nickfargo** [comment]
> Curse of knowledge

Heh, that’s a generous way to put it :)

It’s a good point about providing a classic todo example, if just for the benefit of instant-familiarity. In general I worry about improperly wielding this particular hammer where it isn’t necessarily the best choice — i.e., one does not necessarily benefit from using a **state** library or similar to implement the simpler cases of state machine patterns — as many a perfectly good **state**-less todo app can attest.

But I must admit, that worry is likely misguided; to teach something effectively, I suppose it only makes sense to begin the exposition someplace near the beginning, after all.

Good thoughts, thanks … and many thanks for the help and for feeding back your experiences with it.

**Comment by AdrianRossouw** [comment]
> In general I worry about improperly wielding this particular hammer where it isn’t necessarily the best choice 

If I didn't have a problem that specifically called for a FSM, I would never have stumbled upon state.js.

I'm really not sure if it's because of 'when all you have a screw driver' or the fact that I have it loaded already so I might as well..., but I am finding a lot of places where it turns out I needed a state machine in the first place.

What is an entity such as a backbone.router if not a simple, and not very capable, state machine? [Other people have been learning similar lessons too](http://marksoper.me/Backbone-Routing-Lessons-Learned-Subrouters-and-State-Machines.html), and stuff like ember and angular are all getting routers that are very heavily influenced by state machines.

We'll see in 6 months how I feel about the code I'm writing now =)

**Comment by nickfargo** [comment]
You’re right, certainly, the state pattern is a famous lurker, and eventually seems to reveal itself in all sorts of places.

My restraint comes only from a desire to avoid expressing the fallacy that every application of the pattern is necessarily a good fit for **State** or some similar feature-ful library. Rather in the spirit of, [“Sometimes, the elegant implementation is a function.…”](https://twitter.com/ID_AA_Carmack/status/53512300451201024), as I’m reminded.

But then, restraint can be overused as well. For the sake of crafting helpful examples, at least, it’s ok to wield the hammer liberally.

Routers are a great example class. Presentation of view states is another. Game entities and actors, too … and away we go =)

**Comment by nzfarmer1** [comment]
As a new state user I'm just going to add my 5 cents here.  I found the language a little impenetrable at first.  That said, it was an area that is critical to our work - controlling asynchronous sensor networks.   Having gobbledy gook examples forced me to read your code (and admire your work) slower and in more depth than I would have otherwise.  State machines require rigour.  Many more general APIs simply require copy and paste!

After a few days of nighttime reading I started coding and crafted a working state machine today.  It appears to work fine, thought it would be nice to share it and have some feedback but that's another issue.

Am still struggling a bit with transitions - the events didn't fire when I expected them to.  But I rewrote the code using additional states with various admit, release and enter events.

A general comment here on state and I.o.T.  I am likely not alone here in wondering where all this is going.  Expressing logic for analog controllers/sensors in C++/JavaScript on commodity hardware may seem appealing.  However it can also be daunting.  Ladder logic might have been cumbersome but it provided something of a safe jacket.   In short, writing our gateway handlers with "state" may well prove to be the elusive holy grail we've been looking for.  Thanks for your rigour and generosity in contributing this.

**Comment by nzfarmer1** [comment]
I will say this about the examples.  The one I critically need (transitions) is broken.  A copy and paste failed to compile  for node.js and after tweaking the results were not what I expected.  Am still confused as to how to implement transitions and am wondering whether the library I have is up to date.

**Comment by nickfargo** [comment]
I appreciate the feedback very much. Anything you’d be willing to share of your work would be fantastic;  reviewing specific implementations is always informative and insightful.

**Comment by nzfarmer1** [comment]
This is where I got to.    Extending a simple class/function that works as a relay/actuator handler across a wireless network. It runs as a node-red node - the logic notifies the Remote End Unit, waits for a response confirmation, then notifies the client of the state change.

I like the pattern but no doubt it can be improved.   We used exit events to get around not having transitions.  I still need to add some sort of error on timeout - i.e. the request to change a relay's state has failed.

```
    state( this.constructor.prototype, 'abstract', {
        Off: state(node.params['state'] == 'Off' ? 'default initial':'',{
            admit: 'On,toggleOff',
            events: { enter: function(){
                            msg.topic = node.topic;
                            msg.payload = 'Off';
                            node.params['state'] = msg.payload;
                            node.save();
                }}
        }),
        ToggleOn: {
            admit: 'Off',
            events: { enter: function(){
                           msg.topic = node.match + "/set";
                           msg.payload = 'On';
                           node.send([msg,null]); // Send message to REU to alter physical relay state
                         },exit: function(){ // Alternative to transition. Only possible state is Off. 
                           node.emit('set-complete',node.params);  // Trigger event to update parameters locally and notify client
                    }},
        },
        On: state(node.params['state'] == 'On' ? 'default initial':'',{
            admit: 'Off,toggleOn',
            events: { enter: function(){    
                           node.params['state'] = msg.payload;
                           node.save(); // Save current state
                    }
            }
        }),
        ToggleOff: {
            admit: 'On',
            events: { enter: function(){
                           msg.topic = node.match + "/set";
                           msg.payload = 'Off';
                           node.send([msg,null]); // Send message to REU to alter physical relay state
                        }, exit: function(){ 
                           node.emit('set-complete',node.params); 
                    }},
            },
    });
```

**Comment by nickfargo** [comment]
If I may guess as to the surrounding environment in which the above code appears …

The `this.constructor.prototype` argument suggests to me that this `state()` call is made possibly inside a method, more likely inside of a constructor function … e.g.:

``` js
function AConstructor () {
  state( this.constructor.prototype, 'abstract', { /*...*/ } );
}
```

In any case, this form looks unusual _(see footnote)_. Idiomatic usage applies the state expression either

**(1)** to a prototype, from outside its corresponding constructor:

``` js
function AConstructor () {}
state( AConstructor.prototype, 'abstract', { /*...*/ } );
```

or **(2)** to each instance, from inside a constructor:

``` js
function AConstructor () {
  state( this, 'abstract', { /*...*/ } );
}
```

If your `state()` call is indeed made from inside a constructor, then it may be possible to avoid the #22 bug for now by following pattern **(2)** above, i.e. `state( this, ... )` rather than `state( this.constructor.prototype, ... )`.

---

(What `state( this.constructor.prototype, ... )` suggests is that the prototype is being altered on each invocation of whatever the surrounding function is, which is not proper for any case I can think of. There is the special case of a constructor being used just to create a singleton instance, but even there, it’d still be proper to use `state( this, ... )` inside the constructor.)

**Comment by nzfarmer1** [comment]
Thanks.  The class is intended to be invoked only once.  But using 'this' is better.  Change made.
